openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py

```diff
@@ -0,0 +1,209 @@
+        (3600, 900, 600, False),  # 10 min < 15 min save delay, no save
+        (3600, 900, 1200, False),  # 20 min > 15 min save delay, save occurs
+        (3600, 600, 480, False),  # 8 min < 10 min save delay, no save
+        (3600, 600, 720, False),  # 12 min > 10 min save delay, save occurs
+    )
+    @ddt.unpack
+    @patch("openedx.core.djangoapps.session_inactivity_timeout.middleware.datetime")
+    @patch("openedx.core.djangoapps.session_inactivity_timeout.middleware.auth")
+    @patch("openedx.core.djangoapps.session_inactivity_timeout.middleware.monitoring_utils")
+    def test_process_request_timeout_and_save_behavior(
+        self, timeout_seconds, save_delay_seconds, seconds_elapsed, should_logout,
+        mock_monitoring, mock_auth, mock_datetime
+    ):
+        with override_settings(
+            SESSION_INACTIVITY_TIMEOUT_IN_SECONDS=timeout_seconds,
+            SESSION_ACTIVITY_SAVE_DELAY_SECONDS=save_delay_seconds
+        ):
+            last_touch = datetime(2025, 6, 16, 12, 0, 0)
+            current_time = last_touch + timedelta(seconds=seconds_elapsed)
+            self.request.session[LAST_TOUCH_KEYNAME] = last_touch.isoformat()
+            mock_datetime.utcnow.return_value = current_time
+            mock_datetime.fromisoformat = datetime.fromisoformat
+
+            response = self.middleware.process_request(self.request)  # lint-amnesty, pylint: disable=assignment-from-none
+
+            assert response is None
+
+            if should_logout:
+                assert LAST_TOUCH_KEYNAME not in self.request.session
+                mock_auth.logout.assert_called_once_with(self.request)
+            else:
+                mock_auth.logout.assert_not_called()
+                assert LAST_TOUCH_KEYNAME in self.request.session
+
+                should_save = seconds_elapsed > save_delay_seconds
+
+                if should_save:
+                    assert self.request.session[LAST_TOUCH_KEYNAME] == current_time.isoformat()
+                    mock_monitoring.set_custom_attribute.assert_has_calls([
+                        call("session_inactivity.activity_seen", current_time.isoformat()),
+                        call("session_inactivity.proceed_with_period_save", True),
+                    ], any_order=True)
+                else:
+                    # Session should not be saved, timestamp remains the same
+                    assert self.request.session[LAST_TOUCH_KEYNAME] == last_touch.isoformat()
+                    mock_monitoring.set_custom_attribute.assert_any_call(
+                        "session_inactivity.proceed_with_period_save", False
+                    )
+
+    @ddt.data(
+        # (seconds_elapsed, should_save) - testing around default 900 second save delay
+        (800, False),   # 13.3 min < 15 min default, no save
+        (900, False),   # exactly 15 min default, no save (not exceeded)
+        (1000, True),   # 16.7 min > 15 min default, save occurs
```

Comment:
1. The only difference between testing boundary conditions and not, is the inputs. I think the boundary conditions are more important (e.g. 899,900,901), but if you want both, just have 800,899,900,901,1000 in one test. Note: For testing default setting, you could use abbreviated 899,900,901, and for testing non-default you could use the full list.
2. The code comments in minutes I find needlessly confusing. If you had `# 899 sec < 900 sec default, no save`, it really doesn't matter what this is in minutes, it is easier to match against the 899 and 900 in the test and early comment. This applies elsewhere as well.